python/wikibot/wikibot.py

```diff
@@ -0,0 +1,183 @@
+  sling.CENTURY: pywikibot.WbTime.PRECISION['century'],
+  sling.DECADE: pywikibot.WbTime.PRECISION['decade'],
+  sling.YEAR: pywikibot.WbTime.PRECISION['year'],
+  sling.MONTH: pywikibot.WbTime.PRECISION['month'],
+  sling.DAY: pywikibot.WbTime.PRECISION['day']
+}
+
+class StoreFactsBot:
+  def __init__(self, test=True):
+    self.site = pywikibot.Site("wikidata", "wikidata")
+    self.repo = self.site.data_repository()
+
+    time_str = datetime.datetime.now().isoformat("-")[:19].replace(":","-")
+    if test:
+      record_file_name = "local/data/e/wikibot/test-birth-dates.rec"
+      time_str = "test-" + time_str
+    else:
+      record_file_name = "local/data/e/wikibot/birth-dates.rec"
+    status_file_name = "local/data/e/wikibot/wikibotlog-"+time_str+".rec"
+
+    try:
+      self.record_file = sling.RecordReader(record_file_name)
+      self.status_file = sling.RecordWriter(status_file_name)
+    except :
+      print "Can't open Sling record files", record_file_name, status_file_name
+      sys.exit()
+
+    self.store = sling.Store()
+    self.n_item = self.store["item"]
+    self.n_facts = self.store["facts"]
+    self.n_date_of_birth = self.store["P569"]
+    self.n_provenance = self.store["provenance"]
+    self.n_category = self.store["category"]
+    self.n_method = self.store["method"]
+    self.n_status = self.store["status"]
+    self.n_stored = self.store["stored"]
+    self.n_skipped = self.store["skipped"]
+    self.n_error = self.store["error"]
+    self.n_has_prop = self.store["already has property"]
+    self.n_had_prop = self.store["already had property"]
+    self.n_incons_input = self.store["inconsistent input"]
+    self.n_redirect = self.store["is a redirect page"]
+    self.store.freeze()
+
+    self.source_claim = pywikibot.Claim(self.repo, "P3452") # Inferred from
+    self.time_claim = pywikibot.Claim(self.repo, "P813") # Referenced (on)
+    today = datetime.date.today()
+    time_target = pywikibot.WbTime(year=today.year,
+                                   month=today.month,
+                                   day=today.day)
+    self.time_claim.setTarget(time_target)
+
+  def getSources(self, category):
+    source_target = pywikibot.ItemPage(self.repo, category)
+    self.source_claim.setTarget(source_target)
+    return [self.source_claim, self.time_claim]
+
+  def everHadProp(self, wd_item, prop):
+    # up to 150 revisions covers the full history of 99% of e.g. human items
+    revisions = wd_item.revisions(total=150, content=True)
+    for revision in revisions:
+      try:
+        revision_text = json.loads(revision.text)
+        claims = revision_text['claims']
+      except:
+        pass # unable to extract claims - move to next revision
+      else:
+        if prop in claims: return True
+    return False
+
+  def storeRecords(self, records, batch_size=3):
+    rs = sling.Store(self.store)
+    updated = 0
+    for item, record in records:
+      if updated >= batch_size:
```

Comment:
What about being able to specify a range of records to update? 

You can add flags to control this for test runs, e.g.
```
import sling.flags as flags

flags.define("--first",
             help="first record to update",
             default=0)

flags.define("--last",
             help="last record to update",
             default=sys.maxint)

...
if recno < flags.arg.first or recno > flags.arg.last: continue

if __name__ == '__main__':
  flags.parse()
```

Reply:
SG. Added "first" and "last" cmd-line flags as suggested.